data/transactions/logic/eval.go

```diff
@@ -355,6 +361,10 @@ func EvalStateful(program []byte, params EvalParams) (pass bool, err error) {
 	cx.EvalParams = params
 	cx.runModeFlags = runModeApplication
 	pass, err = eval(program, &cx)
+	if pass && cx.EvalParams.Proto.EnableAppCostPooling {
+		// if eval passes, then budget is always greater than cost, so should not have underflow
+		*cx.EvalParams.PooledApplicationBudget -= cx.cost
```

Comment:
We should be consistent on whether PooledApplicationBudget is sure to be set.  Earlier you checked against nil, here you do not.

Reply:
Good point -- I was wondering if we need to check it at all, but I will add the check here as well.